src/tls/cockpit-certificate-ensure.c

```diff
@@ -72,6 +72,33 @@ read_file (const char     *filename,
   close (fd);
 }
 
+static void
+write_file (int                   dirfd,
+            const char           *dirfd_filename,
+            const char           *filename,
+            const gnutls_datum_t *data,
+            uid_t                 uid,
+            gid_t                 gid)
+{
+  /* Just open the file directly: it doesn't exist yet and nobody will
+   * look at it until after we're done here.
+   */
+  int fd = openat (dirfd, filename, O_CREAT | O_EXCL | O_WRONLY, 0400);
+
+  if (fd == -1)
+    err (EXIT_FAILURE, "%s/%s: creat", dirfd_filename, filename);
+
+  if (fchown (fd, uid, gid) != 0)
+    err (EXIT_FAILURE, "%s/%s: fchown", dirfd_filename, filename);
```

Comment:
IMHO it would be nice if c-tls would not even be able to write the file. How about only setting the group to gid and making the file 640?

Also, for robustness the fchown should happen after write() was complete.

Reply:
Ack -- just owning it to cockpit-ws:root would also be sufficient, but indeed all of this makes fairly little difference. cockpit-tls must be able to write the directory to unlink the files (and I like that). So feel free to just keep it as-is, it was mostly thinking aloud.